In [1]:
# 📦 Импорт библиотек
import os
import sys
from pathlib import Path

import pandas as pd

# 🔍 EDA (первичный обзор данных)

**Цель:** проверить данные на пропущенные значения, их форматирование и выбросы

**Автор:** Антон Лысогор

**Дата:** 2026-05-01

## Загрузка данных

In [2]:
# Проверка, что используется именно окружение проекта
# Путь должен быть внури проекта ~/ml-hackathon/.venv/bin/python
print(f"Путь к интерпретатору: {sys.executable}")

Путь к интерпретатору: /home/itj/ml-hackathon/.venv/bin/python


In [3]:
# 🗄️ Загрузка данных

#  Адаптация путей под проект
#
# Всегда вычисляем корень относительно ПАПКИ ноутбука,
# а не относительно того, где мы сейчас «гуляем» через chdir
# Если ноутбук в notebooks/eda, то корень — это два уровня вверх
ROOT = Path(os.getcwd())

# Проверка: если мы уже в корне или ушли не туда,
# этот блок поможет не подняться выше проекта (ml-hackathon)
if "notebooks" in str(ROOT):
    ROOT = ROOT.parent.parent

DATA_ROOT = ROOT / "data" / "train"

print(f"Рабочий корень ROOT: {ROOT}")
print(f"Файлы будут искаться в DATA_ROOT: {DATA_ROOT}")

# DATA_ROOT = Path('data/train')
user = pd.read_csv(DATA_ROOT / "user.csv")
shift = pd.read_csv(DATA_ROOT / "shift.csv")
event = pd.read_csv(DATA_ROOT / "event.csv")

print(f"✅ Loaded: users={len(user)}, shifts={len(shift)}, events={len(event)}")

Рабочий корень ROOT: /home/itj/ml-hackathon
Файлы будут искаться в DATA_ROOT: /home/itj/ml-hackathon/data/train
✅ Loaded: users=8802, shifts=65624, events=840850


## Первичный обзор данных

In [4]:
user

,location_id,is_strict_location,id,has_mk
0,32,False,2223888b9607301686f0f19b856aa218,True
1,254,False,401f7d0cf994af0397966abb6e73c2c7,True
2,234,False,02c8966fce79f14cf65f54d300678995,True
3,242,False,5b896a6d98134a1bdc7a80cf74841e01,True
4,38,False,eb3b9ea10eaf6877f83d78cfd947a3b8,True
...,...,...,...,...
8797,432,True,4b6d60fd65dccec5f1899cfdd35433d2,True
8798,5024,True,8720ea49a678792eead5153efa0d91e6,True
8799,1885,True,6183a2da880a2d6d27c1d2c144765098,False
8800,3947,True,69e8ddee628473fc787d2ac159ebddb7,True


In [5]:
user.info()

<class 'pandas.DataFrame'>
RangeIndex: 8802 entries, 0 to 8801
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   location_id         8802 non-null   int64
 1   is_strict_location  8802 non-null   bool 
 2   id                  8802 non-null   str  
 3   has_mk              8802 non-null   bool 
dtypes: bool(2), int64(1), str(1)
memory usage: 154.9 KB


In [6]:
summary = pd.DataFrame({"dtype": user.dtypes, "unique_count": user.nunique()})
print(summary)

                    dtype  unique_count
location_id         int64          4772
is_strict_location   bool             2
id                    str          8802
has_mk               bool             2


Таблица ***user*** — всё нормально, пропущенных нет, поле *id* — все значения уникальны, числовые и булевые значения в соответствующих форматах.

In [7]:
shift

,id,start_at,location_id,task_type,employer_id,workplace_id,need_mk,id_differential,hours,reward,capacity
0,57080,2026-01-01 07:00:00+00:00,234,Выкладка товара,02d2b372ccc7404c03bf705d67360723,67cffef2e08cf63c214fc0b9f6273cd5,True,False,12,2800.0,1
1,59080,2026-01-01 07:00:00+00:00,41,Уборка помещений,6cda58e6bf56ba0ddde0c84212300104,48c8cb23493ca2e467a207d840cdb59e,True,False,12,2500.0,5
2,32676,2026-01-01 07:00:00+00:00,234,Выкладка товара,02d2b372ccc7404c03bf705d67360723,6d8a2900d0cd692005b919959d9f3e2b,True,False,12,3700.0,1
3,37968,2026-01-01 08:00:00+00:00,166,Выкладка товара,fb3d1aaa7ddb54fd11604d5529a7774b,22e8cc4635d2e9ca20095842983559aa,True,False,15,5750.0,1
4,9587,2026-01-01 08:00:00+00:00,234,Выкладка товара,02d2b372ccc7404c03bf705d67360723,7e17fba7df94be7f98948340eee1935c,True,False,12,1300.0,1
...,...,...,...,...,...,...,...,...,...,...,...
65619,83619,2026-03-22 23:00:00+00:00,234,Выкладка товара,02d2b372ccc7404c03bf705d67360723,9d3164ba1c88ced2fe41615f66064110,True,False,9,2800.0,1
65620,83643,2026-03-22 23:00:00+00:00,234,Выкладка товара,02d2b372ccc7404c03bf705d67360723,b2baf1b3e609429e59fe07eafaf21ac7,True,False,10,4500.0,1
65621,83644,2026-03-22 23:00:00+00:00,32,Мойка посуды и инвентаря,02d2b372ccc7404c03bf705d67360723,f2ac9c5b0557aadb8783b3f802aa2843,True,False,8,3400.0,1
65622,83646,2026-03-22 23:00:00+00:00,234,Мойка посуды и инвентаря,02d2b372ccc7404c03bf705d67360723,bdd0403d79e3436bbbb8998d807b8c33,True,False,9,5250.0,1


In [8]:
shift.info()

<class 'pandas.DataFrame'>
RangeIndex: 65624 entries, 0 to 65623
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               65624 non-null  int64  
 1   start_at         65624 non-null  str    
 2   location_id      65624 non-null  int64  
 3   task_type        65624 non-null  str    
 4   employer_id      65624 non-null  str    
 5   workplace_id     65624 non-null  str    
 6   need_mk          65624 non-null  bool   
 7   id_differential  65624 non-null  bool   
 8   hours            65624 non-null  int64  
 9   reward           65624 non-null  float64
 10  capacity         65624 non-null  int64  
dtypes: bool(2), float64(1), int64(4), str(4)
memory usage: 4.6 MB


In [9]:
df = shift
summary = pd.DataFrame({"dtype": df.dtypes, "unique_count": df.nunique()})
print(summary)

                   dtype  unique_count
id                 int64         65624
start_at             str          2721
location_id        int64           275
task_type            str            29
employer_id          str            64
workplace_id         str          2417
need_mk             bool             2
id_differential     bool             2
hours              int64            23
reward           float64            89
capacity           int64            10


In [10]:
shift.describe()

,id,location_id,hours,reward,capacity
count,65624.000000,65624.000000,65624.000000,65624.000000,65624.000000
mean,66326.723653,174.279913,9.851335,2946.819761,1.182860
std,31167.222341,157.432103,2.769535,1536.004042,0.702352
min,4.000000,1.000000,1.000000,1000.000000,1.000000
25%,41998.250000,118.000000,7.000000,1900.000000,1.000000
50%,73161.500000,201.000000,11.000000,2700.000000,1.000000
75%,91929.250000,234.000000,12.000000,3700.000000,1.000000
max,112841.000000,5782.000000,23.000000,22000.000000,10.000000


In [11]:
shift.describe(include=str)

,start_at,task_type,employer_id,workplace_id
count,65624,65624,65624,65624
unique,2721,29,64,2417
top,2026-01-07 08:00:00+00:00,Выкладка товара,fb3d1aaa7ddb54fd11604d5529a7774b,67cffef2e08cf63c214fc0b9f6273cd5
freq,227,28161,20326,891


Таблица ***shift***, или *рабочая смена*. Пропущенных значений нет, все id уникальны.

* *start_at* - начало смены,  время в строковой переменной, надо переводить во время или убедиться, что оно в дальнейшем переводится в него.
* 275 локаций, 29 типов работ, 64 заказчика.
* *workplace_id*: Идентификатор конкретного рабочего места внутри локации. **Примечение:** пока непонятно, что значит.
* *need_mk* — начличие медицинской книжки, связано с полем *has_mk* в таблице *user*.
* *id_differential* — машинный перевод: "Вероятно, признак того, отличается ли оплата в этой смене от стандартной (например, повышенная ставка)." **Примечение:** пока непонятно, что значит.
* *hours* — длительность смены, от 1 часа до 23, средняя 10 часов с отклонением в 3 часа, медианное значение 11 часов.
* *reward* — вознаграждение за смену от 1000 до 22000. Середина и среднее в районе 3000. Из этого поля можем получить расчётную ставку за один час. Чем она больше, тем больше пользователь будет мотивирован выйти на смену.
* *capacity* — количество человек на смену. Обычно 1, максимум 10.

In [12]:
event

,id,shift_id,user_id,interaction,ts
0,3e0c68e79a79f3cfbb5dc33a94539538,64679,d0b587858e4eba68a2d5b71dd7cc91b3,VIEW,2026-02-23
1,4219556E8C93233128D8218577DD9F06,64679,d0b587858e4eba68a2d5b71dd7cc91b3,APPLY,2026-02-23
2,0BF1B8CD6627920077C54DEA87713AA5,64679,d0b587858e4eba68a2d5b71dd7cc91b3,FINISHED,2026-02-23
3,65664e85b7d79b5f370ea5db40b40cef,65231,d0b587858e4eba68a2d5b71dd7cc91b3,VIEW,2026-02-24
4,959BFA3BE2079E32FD063B60F7251419,65231,d0b587858e4eba68a2d5b71dd7cc91b3,APPLY,2026-02-24
...,...,...,...,...,...
840845,31E6CFE43FD431FBCF8B82921B54158A,104572,27164755878b390657e9374fe4e444c9,SYSTEM_CANCEL,2026-03-16
840846,1399572f1976913673d534c425d56aa0,104851,27164755878b390657e9374fe4e444c9,VIEW,2026-03-20
840847,22DE89AC51704DAEA2700AC78A07FC7B,104851,27164755878b390657e9374fe4e444c9,APPLY,2026-03-20
840848,54BA1E7B4DD6080A0E7D9884AD83678D,104851,27164755878b390657e9374fe4e444c9,SYSTEM_CANCEL,2026-03-21


In [13]:
event.info()

<class 'pandas.DataFrame'>
RangeIndex: 840850 entries, 0 to 840849
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype
---  ------       --------------   -----
 0   id           840850 non-null  str  
 1   shift_id     840850 non-null  int64
 2   user_id      840850 non-null  str  
 3   interaction  840850 non-null  str  
 4   ts           840850 non-null  str  
dtypes: int64(1), str(4)
memory usage: 32.1 MB


In [17]:
event.shape

(840850, 5)

In [14]:
df = event
summary = pd.DataFrame({"dtype": df.dtypes, "unique_count": df.nunique()})
print(summary)

             dtype  unique_count
id             str        421003
shift_id     int64         65624
user_id        str          8794
interaction    str             5
ts             str            81


In [15]:
event.interaction.value_counts()

interaction
VIEW             767065
APPLY             36897
FINISHED          15910
USER_CANCEL       11889
SYSTEM_CANCEL      9089
Name: count, dtype: int64

In [16]:
event.ts.value_counts()

ts
2026-02-28    16650
2026-03-01    15301
2026-02-27    15259
2026-01-03    14891
2026-01-04    14814
              ...  
2026-02-23     7100
2026-02-18     6930
2026-03-20     6705
2026-03-21     4059
2026-03-22     1367
Name: count, Length: 81, dtype: int64

Таблица ***event***, пожалуй, самая важная. по видимому представляет лог действий. Пользователь, посмотрел VIEW, при этом может смотреть несколько раз; записался  APPLY и либо вышел FINISHED, либо отказался USER_CANCEL. Ещё есть вариант SYSTEM_CANCEL — смена была отменена.

Пропущенных значений нет.

* *shift_id* — 65624 в таблице *shift* столько же.
* *user_id* — 8794 и в таблице *user* их 8802. Есть данные о действиях почти о всех пользователях.
* *ts*, time stamp - в строковом виде.

## Вывод

В целом все данные корректны, только дата тербует преобразования. 

**Дальнейшие шаги:** перед преобразованием и наращиванием фич, теперь надо изучить baseline, что конкретно там уже сделано.

# 🔎 Углубленный анализ

**Цель:** попробовать данные, выявить устойчивые паттерны, проверить как их обрабатывает модель

**Автор:** Антон Лысогор

**Дата:** 2026-05-09